### PR TITLE
Bugfix/early stopping

### LINKED ISSUE #63: Early stopping with patience=1 shouldn't stop in the first epoch
cc @skeydan

### LINKED ISSUE #64: Combining early stopping with csv logger callback, metrics for last epoch don't get logged
To be precise, the code combines
- min_epoch
- early stopping
- csv logger

Code:

```
fitted <- convnet %>%
  setup(
    loss = nn_bce_with_logits_loss(),
    optimizer = optim_adam,
    metrics = list(
      luz_metric_binary_accuracy_with_logits()
    )
  ) %>%
  fit(train_dl, epochs = c(5,10), valid_data = valid_dl,
      callbacks = list(luz_callback_early_stopping(), luz_callback_csv_logger("logs.csv")),
      verbose = TRUE)
```

CSV:

```
1,"train",0.67555393654533,0.559885714285714
1,"valid",0.643152226962006,0.632266666666667
2,"train",0.613922604913032,0.658685714285714
2,"valid",0.567475510558594,0.706666666666667
3,"train",0.571434665587092,0.701885714285714
3,"valid",0.518853313394828,0.740933333333333
4,"train",0.537284926724172,0.731257142857143
4,"valid",0.481547881831238,0.769066666666667
```

Console:

```
Train metrics: Loss: 0.5081 - Acc: 0.7493                                              
Valid metrics: Loss: 0.4434 - Acc: 0.7968
Early stopping at epoch 5 of 10
```

### PR DESCRIPTION
**Pull request by dfalbel**
This PR fixes 2 issues:

1. early stopping would always happen in the first epoch if patience = 1 because we would compare the metric with itself- meaning that there was no improvement and then we stopped. This now fixed by simply skipping the early stopping logic when in the first epoch. Fix #63 

2. We can now control the `weight` of a callback, ie the order it should be called in the callback list. Callbacks with no weight attribute are called in the same order they are specified. Early stopping is now a callback with `weight = Inf` meaning that its the last callback to be called. Fix #64 

We also moved much of the code for the early stopping callback to a different file, so it's easier to test and navigate.

### CODE REVIEW

**Review by skeydan — APPROVED** [review]
Looks great to me! I really like the idea with the weights. One very minor suggestion, what do you think of renaming `weight` to `priority` to have a more descriptive variable name?

**Comment by dfalbel** [comment]
I think the problem with 'priority' is that then high-priorities must be called first, I guess? Since probably the more common use case is to postpone a callback call it will be harder to teach that it should have negative priorities.

**Review by skeydan — COMMENTED** [review]
I know this is merged, but still noting the typos in case you want to update (I mean, what follows is the corrected version, to copy-paste as a whole:

By default, callbacks are called in the same order as they were passed to `fit` (`predict` or `evaluate`), but you can provide a `weight` attribute that will control the order that it will be called. For example if a callback has `weight = 10` and the other has `weight = 1` then the first one is called after the second one.
Callbacks that don't specify a `weight` attribute are considered `weight = 0`. A few built-in callbacks in luz already provide a weight value, for example the `?luz_early_stopping_callback`, since in general we want to run it as the last thing in the loop.

**Comment by skeydan** [comment]
> I think the problem with 'priority' is that then high-priorities must be called first, I guess? 

Maybe, not really sure actually ... priorities tend to be weird anyway, in common usage, or in Linux speak ... ;-))

It's just that weight does not sound as if it was about chronological order, but importance (which it is not). Anyway, I can't think of a good alternative right now, `position` being ambiguous (and probably, is mostly understood as referring to things spatial), and it's not really "high prio" ;-))
